### ISSUE TITLE
Error installing PyMQI via pip

### ISSUE THREAD

**Issue by peterboldizs** [dsuch/pymqi#15]
I have tried to install pymqi using pip. 

mqm@mqserver:~$ sudo pip install pymqi
Downloading/unpacking pymqi
  Downloading pymqi-1.4.tar.gz (64kB): 64kB downloaded
  Running setup.py (path:/tmp/pip_build_root/pymqi/setup.py) egg_info for package pymqi
    Building PyMQI client 64bits
    
    warning: no files found matching 'scripts/*.py'
Installing collected packages: pymqi
  Running setup.py install for pymqi
    Building PyMQI client 64bits
    building 'pymqe' extension
    x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -DPYQMI_SERVERBUILD=0 -I/opt/mqm/inc -I/usr/include/python2.7 -c pymqe.c -o build/temp.linux-x86_64-2.7/pymqe.o
    pymqe.c:78:18: fatal error: cmqc.h: No such file or directory
     `#include <cmqc.h>`
                      ^
    compilation terminated.
    error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
    Complete output from command /usr/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip_build_root/pymqi/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-CJY6fY-record/install-record.txt --single-version-externally-managed --compile:
    Building PyMQI client 64bits

running install

running build

running build_py

creating build

creating build/lib.linux-x86_64-2.7

copying pymqi.py -> build/lib.linux-x86_64-2.7

copying CMQC.py -> build/lib.linux-x86_64-2.7

copying CMQCFC.py -> build/lib.linux-x86_64-2.7

copying CMQXC.py -> build/lib.linux-x86_64-2.7

copying CMQZC.py -> build/lib.linux-x86_64-2.7

running build_ext

building 'pymqe' extension

creating build/temp.linux-x86_64-2.7

x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -DPYQMI_SERVERBUILD=0 -I/opt/mqm/inc -I/usr/include/python2.7 -c pymqe.c -o build/temp.linux-x86_64-2.7/pymqe.o

pymqe.c:78:18: fatal error: cmqc.h: No such file or directory

 `#include <cmqc.h>`
                  ^

compilation terminated.

error: command 'x86_64-linux-gnu-gcc' failed with exit status 1

----------------------------------------
Cleaning up...
Command /usr/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip_build_root/pymqi/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-CJY6fY-record/install-record.txt --single-version-externally-managed --compile failed with error code 1 in /tmp/pip_build_root/pymqi
Storing debug log for failure in /home/mqm/.pip/pip.log

Could you please help?
I am using Ubuntu Server 14.04
Linux mqserver 3.16.0-30-generic #40~14.04.1-Ubuntu SMP Thu Jan 15 17:43:14 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

**Comment by robholmes** [comment]
I have the same issue and the only reason I can see is that `cmqc.h` doesn't exist! It feels to me like there is a dependency missing, but I have no idea why or what it is.

@dsuch Have you got any idea what we're missing?

**Comment by robholmes** [comment]
@peterboldizs I have found the solution, which really should be documented somewhere, but maybe we should of magically known this... We have to install the Websphere MQ Client and SDK!

#### Download from IBM

1. Goto http://www-01.ibm.com/software/integration/wmq/clients/ and click the version you need (*for me it was WebSphere MQ V7.5 Clients*)
2. Scroll down to **Download package** and click HTTPS download link for the latest version
3. Select only the install image for your operating system, and click continue
4. Follow the further instructions until you have the downloaded archive

#### Install

1. Extract the archive to a directory of your choice (only temporary, you can delete later)
2. Open a terminal and `cd` to the directory from above
3. Run `sudo ./mqlicense.sh -accept` to accept the license
4. Install `rpm` if you don't have it already: `sudo apt-get install rpm`
5. Install Runtime: `sudo rpm --prefix /opt/mqm -ivh --nodeps --force-debian MQSeriesRuntime-7.5.0-5.x86_64.rpm`
6. Install Client: `sudo rpm --prefix /opt/mqm -ivh --nodeps --force-debian MQSeriesClient-7.5.0-5.x86_64.rpm`
7. Install SDK: `sudo rpm --prefix /opt/mqm -ivh --nodeps --force-debian MQSeriesSDK-7.5.0-5.x86_64.rpm`

#### Summary

You should now be able to run `sudo pip install pymqi` without error (you should see the output of "Successfully installed pymqi"). This works because the setup.py script looks for includes and libraries in the /opt/mqm/inc and /opt/mqm/lib* directories.

Hope this helps some others get around the problem, I've wasted at least a day on this now!

**Comment by dsuch** [comment]
Hello everyone,

yes, PyMQI is a wrapper around WMQ C libraries rather than a standalone pure-Python package.

Actually, I'd prefer for PyMQI to be pure-Python but WMQ protocols are proprietary and undocumented so there is slim chance it will happen soon.

That all said, yes, you need to install WMQ first, otherwise the C parts of PyMQI cannot find the C headers to compile against.

I apologize for having caused trouble, it just honestly never occurred to me that this fact needs to be mentioned anywhere. Obviously, this is my fault.

How would you recommend this could be amended? Would adding a prominent paragraph somewhere in the installation docs do the trick?

Thanks a lot!

**Comment by peterboldizs** [comment]
Greetings, 
Thanks @robholmes  for the detailed explanation.
I have followed the instructions given by Rob and installed the MQ client packages to my development box. Afterwards, I have tried to run sudo pip install pymqi. This has given the following output:

Downloading/unpacking pymqi
  Downloading pymqi-1.4.tar.gz (64kB): 64kB downloaded
  Running setup.py (path:/tmp/pip_build_root/pymqi/setup.py) egg_info for package pymqi
    Building PyMQI client 64bits
    
    warning: no files found matching 'scripts/*.py'
Installing collected packages: pymqi
  Running setup.py install for pymqi
    Building PyMQI client 64bits
    building 'pymqe' extension
    x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -DPYQMI_SERVERBUILD=0 -I/opt/mqm/inc -I/usr/include/python2.7 -c pymqe.c -o build/temp.linux-x86_64-2.7/pymqe.o
    pymqe.c:108:20: fatal error: Python.h: No such file or directory
     #include "Python.h"
                        ^
    compilation terminated.
    error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
    Complete output from command /usr/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip_build_root/pymqi/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-buwPgm-record/install-record.txt --single-version-externally-managed --compile:
    Building PyMQI client 64bits

running install

running build

running build_py

creating build

creating build/lib.linux-x86_64-2.7

copying pymqi.py -> build/lib.linux-x86_64-2.7

copying CMQC.py -> build/lib.linux-x86_64-2.7

copying CMQCFC.py -> build/lib.linux-x86_64-2.7

copying CMQXC.py -> build/lib.linux-x86_64-2.7

copying CMQZC.py -> build/lib.linux-x86_64-2.7

running build_ext

building 'pymqe' extension

creating build/temp.linux-x86_64-2.7

x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -DPYQMI_SERVERBUILD=0 -I/opt/mqm/inc -I/usr/include/python2.7 -c pymqe.c -o build/temp.linux-x86_64-2.7/pymqe.o

pymqe.c:108:20: fatal error: Python.h: No such file or directory

 #include "Python.h"

                    ^

compilation terminated.

error: command 'x86_64-linux-gnu-gcc' failed with exit status 1

----------------------------------------
Cleaning up...
Command /usr/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip_build_root/pymqi/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-buwPgm-record/install-record.txt --single-version-externally-managed --compile failed with error code 1 in /tmp/pip_build_root/pymqi
Storing debug log for failure in /home/mqm/.pip/pip.log

I have also tried to run the above on a box where a full MQ installation exists (it has several queues, queue managers etc.). Unfortunately it has the same results.

**Comment by dsuch** [comment]
@peterboldizs - please install the ```python-dev``` package using yum or apt-get (I'm not sure what your OS is), this is where Python.h is.

**Comment by robholmes** [comment]
Hi @peterboldizs, I'm sorry I forgot to mention about installing `python-dev`, I've updated my instructions to cover that now.

**Comment by peterboldizs** [comment]
I have installed python-dev and now it works: I have successfully installed PyMQI.
Thanks @dsuch and @robholmes for your support!

**Comment by dsuch** [comment]
Closing the ticket - thanks everyone!

**Comment by mritchie712** [comment]
Does anyone know how to get this running on Mac OS X?

**Comment by dsuch** [comment]
Hi @mritchie712 - PyMQI is a wrapper around WebSphere MQ libraries. First question is if IBM offers the libraries for OS X?

**Comment by mritchie712** [comment]
No, I was working on a way to get it to but have failed miserably, going with this instead:

https://www.ibm.com/developerworks/community/blogs/messaging/entry/introducing_a_docker_image_for_mq_advanced_for_developers?lang=en

**Comment by dsuch** [comment]
@mritchie712 Aha, OK, per https://github.com/ibm-messaging/mq-docker/blob/master/8.0.0/Dockerfile this is Ubuntu then under Docker so the answer is yes, you can certainly compile PyMQI under Ubuntu.

Cheers.

**Comment by dsuch** [comment]
Hello @nrai-bagi - I am not able to provide free support on systems other than Linux. Feel free to [contact me directly](https://dsuch.github.io/pymqi/support.html) if you are interested in commercial support.

Otherwise, this is an open-source library so you can look its details in the source code or you can wait for someone in the community to possibly help you out.

**Comment by toppgun** [comment]
i have posted same query on /issuse/35 but looks like it is closed.
Hello @robholmes @peterboldizs @dsuch 

I am trying to connect to ibm mq using python on linux
i have latest python 3.8 on linux and IBM 7.5
i have installed pymqi-1.9.3
i have the libraries as below files in directory /python/lib/python3.8/site-packages/pymqi-1.9.3-py3.8-linux-x86_64.egg/pymqi>
CMQC.py
CMQZC.py
CMQXC.py
CMQCFC.py
init.py
pymqe.py
pymqe.cpython-38-x86_64-linux-gnu.so
pycache

however, i get the below error when i use import pymqi in python

Traceback (most recent call last):
File "/python/lib/python3.8/site-packages/pymqi-1.9.3-py3.8-linux-x86_64.egg/pymqi/init.py", line 110, in
from . import pymqe
ImportError: libmqic_r.so: cannot open shared object file: No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "abc_mq.py", line 2, in
import pymqi
File "/python/lib/python3.8/site-packages/pymqi-1.9.3-py3.8-linux-x86_64.egg/pymqi/init.py", line 112, in
import pymqe # Backward compatibility
ModuleNotFoundError: No module named 'pymqe'

Any pointer will be really helpful

**Comment by JoshMcMQ** [comment]
Why are you adding comments to random closed issue that do not appear even related to the issue you are reporting?  It would be better to open a new issue if you have searched Google, this repository,  and stackoverflow and could not find a solution. 

That being said I will provide one comment here if it doesn't help and you have exhausted your own efforts in resolving it,  then please open a new issue. 

---

If you have installed a full MQ client install make sure you do one of the following:

1. Set the install as the primary install. 
2. Run the `setmqenv` command under the `bin` directory under MQ install location so that variables like `LD_LIBRARY_PATH` are correctly set to point to the MQ libraries for the MQ install. 
3. Manually export `LD_LIBRARY_PATH` to point to the `lib` (32bit) or `lib64` (64bit) directory under the MQ install location. 

If you have extracted the redistributable client then do one of the following:
1. Run the `setmqenv` command under the `bin` directory where you extracted the files so that variables like `LD_LIBRARY_PATH` are correctly set to point to the MQ libraries extracted there. 
2. Manually export `LD_LIBRARY_PATH` to point to the lib (32bit) or lib64 (64bit) directory under the directory where you extracted the files.

**Comment by sastorsl** [comment]
Since it took _way_ too long for me to see the comment from @robholmes - and only after I found the answer myself I'm dropping this at the bottom for the _next_ person:

Basically, as Rob says, you need the MQSeriesSDK as well.

```bash
MQSeriesGSKit-8.0.0-2.x86_64
MQSeriesJRE-8.0.0-2.x86_64
MQSeriesClient-8.0.0-2.x86_64
MQSeriesRuntime-8.0.0-2.x86_64
MQSeriesSDK-8.0.0-2.x86_64
```

To add something of value:
I had to install this on an old RedHat 6 server (RHEL6) - hence the old MQ version... (glibc issues)
So I also had to install Python-27

```bash
python27-2.7
python27-devel  # To get Python.h
```

Since pymqi was going to be set up in a controlled environment I installed pymqi as a separate user:

```bash
python2.7 setup.py install --prefix=$HOME
```

**Comment by 776166** [comment]
IBM MQ — has the whorst installation process, manual and site I ever seen.
